syntax = "proto3";
package github.com.dnswlt.hexz;

import "google/protobuf/timestamp.proto";
option go_package = "github.com/dnswlt/hexz/hexzpb";

message Board {
    enum GameState {
        UNSPECIFIED = 0;
        INITIAL = 1;
        RUNNING = 2;
        FINISHED = 3;
    }
    int32 turn = 1;
    int32 move = 2;
    int32 last_revealed = 3;
    repeated Field flat_fields = 4;
    repeated int32 score = 5;
    repeated ResourceInfo resources = 6;
    GameState state = 7;
}

message Field {
    enum CellType {
        UNSPECIFIED = 0;
        NORMAL = 1;
        DEAD = 2;
        GRASS = 3;
        ROCK = 4;
        FIRE = 5;
        FLAG = 6;
        PEST = 7;
        DEATH = 8;
    }
    CellType type = 1;
    int32 owner = 2;
    bool hidden = 3;
    int32 value = 4;
    int32 blocked = 5;
    int32 lifetime = 6;
    repeated int32 next_val = 7;
}

message ResourceInfo {
    // Has exactly as many elements as there are cell types.
    repeated int32 num_pieces = 1;
}

// This message contains all data to restore a game state.
// It can be used to save and load games in a memory store.
message GameState {
    google.protobuf.Timestamp created = 1;
    // Use oneof with one element per game engine since proto3 does not support extensions.
    oneof engine_state {
        GameEngineFlagzState flagz = 2;
    }
}

// The encoded state of a GameEngineFlagz. Used for saving and loading games.
message GameEngineFlagzState {
    Board board = 1;
    int32 free_cells = 2;
    // Always exactly two elements, one per player.
    repeated int32 normal_moves = 3;
    // History of moves made so far.
    repeated GameEngineMove moves = 4;
}

message GameEngineMove {
    int32 player_num = 1;
	int32 move = 2;
	int32 row = 3;
	int32 col = 4;
	Field.CellType cell_type = 5;
}
